src/core/common/timer.hpp

```diff
@@ -62,56 +63,224 @@ class Timer;
  */
 
 /**
- * This class implements the timer scheduler.
+ * This class implements a timer.
  *
  */
-class TimerScheduler
+class TimerBase: public Ip6Locator, public Context
 {
-    friend class Timer;
+    friend class TimerSchedulerBase;
 
 public:
+
+    enum
+    {
+        kMaxDt = (1UL << 31) - 1,  //< Maximum permitted value for parameter `aDt` in `Start` and `StartAt` method.
+    };
+
     /**
-     * This constructor initializes the object.
+     * This function pointer is called when the timer expires.
+     *
+     * @param[in]  aTimer    A reference to the expired timer instance.
      *
      */
-    TimerScheduler(void);
+    typedef void (*Handler)(TimerBase &aTimer);
 
     /**
-     * This method adds a timer instance to the timer scheduler.
+     * This constructor creates a timer instance.
      *
-     * @param[in]  aTimer  A reference to the timer instance.
+     * @param[in]  aIp6        A reference to the IPv6 network object.
+     * @param[in]  aHandler    A pointer to a function that is called when the timer expires.
+     * @param[in]  aContext    A pointer to arbitrary context information.
      *
      */
-    void Add(Timer &aTimer);
+    TimerBase(Ip6::Ip6 &aIp6, Handler aHandler, void *aContext):
+        Ip6Locator(aIp6),
+        Context(aContext),
+        mHandler(aHandler),
+        mFireTime(0),
+        mNext(this) {
+    }
 
     /**
-     * This method removes a timer instance to the timer scheduler.
+     * This method returns the fire time of the timer.
      *
-     * @param[in]  aTimer  A reference to the timer instance.
+     * @returns The fire time in milliseconds.
      *
      */
-    void Remove(Timer &aTimer);
+    uint32_t GetFireTime(void) const { return mFireTime; }
 
     /**
-     * This method processes the running timers.
+     * This method indicates whether or not the timer instance is running.
+     *
+     * @retval TRUE   If the timer is running.
+     * @retval FALSE  If the timer is not running.
      *
      */
-    void ProcessTimers(void);
+    bool IsRunning(void) const { return (mNext != this); }
+
+protected:
+    /**
+     * This method indicates if the fire time of this timer is strictly before the fire time of a second given timer.
+     *
+     * @param[in]  aTimer   A reference to the second timer object.
+     * @param[in]  aNow     The current time (may in milliseconds or microsecond, which depends on the timer type).
+     *
+     * @retval TRUE  If the fire time of this timer object is strictly before aTimer's fire time
+     * @retval FALSE If the fire time of this timer object is the same or after aTimer's fire time.
+     *
+     */
+    bool DoesFireBefore(const TimerBase &aTimer, uint32_t aNow);
+
+    void Fired(void) { mHandler(*this); }
+
+    Handler         mHandler;
+    uint32_t        mFireTime;
+    TimerBase      *mNext;
+};
+
+/**
+ * This class implements the millisecond timer.
+ *
+ */
+class TimerMilli: public TimerBase
```

Comment:
We could probably rename `TimerBase` to just `Timer`, given that we already have more-specific `TimerMilli` and `TimerMicro`.